utils/stall_free.hh

```diff
@@ -56,14 +58,93 @@ void merge_to_gently(std::list<T>& list1, const std::list<T>& list2, Compare com
     }
 }
 
-template<class T>
-seastar::future<> clear_gently(std::list<T>& list) {
-    return repeat([&list] () mutable {
-        if (list.empty()) {
-            return seastar::stop_iteration::yes;
-        }
-        list.pop_back();
-        return seastar::stop_iteration::no;
+template <typename T>
+concept GentlyClearable = requires (T x) {
+    { x.clear_gently() } -> std::same_as<future<>>;
+};
+
+template <typename T>
+concept SmartPointer = requires (T x) {
+    { x.reset() } -> std::same_as<void>;
+    { x.get() } -> std::same_as<typename T::element_type*>;
+};
+
+template <typename T>
+concept VectorLike = requires (T x, size_t n) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.back() } -> std::same_as<typename T::value_type&>;
+    { x.pop_back() } -> std::same_as<void>;
+};
+
+template <typename T>
+concept Container = !VectorLike<T> && requires (T x) {
+    { x.empty() } -> std::same_as<bool>;
+    { x.erase(x.begin()) } -> std::same_as<typename T::iterator>;
+};
+
+template <typename T, std::size_t N>
+future<> clear_gently(std::array<T, N>&a) noexcept;
+
+template <Container T>
+future<> clear_gently(T& c) noexcept;
+
+template <GentlyClearable T>
+future<> clear_gently(T& o) noexcept(noexcept(o.clear_gently())) {
+    return o.clear_gently();
+}
+
+template <typename T>
+future<> clear_gently(T&) noexcept {
+    return make_ready_future<>();
+}
```

Comment:
Why do we want such a do-nothing implementation? Isn't it dangerous that someone can call clear_gently(x) and it will do absolutely nothing?

Reply:
It's defined for internal use to stop recursion when we get to leaf objects that are not gently clearable.
Initially I defined `maybe_clear_gently` for the intenal calls and that could be defined in an internal namespace to make it clear that it's not part of the user api.  Same goes for calling clear gently on a mapped `value_type` that is a `std:pair<Key, value>`.

Reply:
Fixed in v6 (implemented in `utils::internal` namespace)